lib/vector-core/src/stream/batcher/mod.rs

```diff
@@ -0,0 +1,197 @@
+pub mod config;
+pub mod data;
+pub mod limiter;
+
+use futures::stream::{Fuse, Stream};
+use futures::{Future, StreamExt};
+use pin_project::pin_project;
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use tokio::time::Sleep;
+
+use config::BatchConfig;
+
+#[pin_project]
+pub struct Batcher<S, C> {
+    config: C,
+
+    #[pin]
+    /// The stream this `Batcher` wraps
+    stream: Fuse<S>,
+
+    #[pin]
+    timer: Maybe<Sleep>,
+}
+
+/// An `Option`, but with pin projection
+#[pin_project(project = MaybeProj)]
+pub enum Maybe<T> {
+    Some(#[pin] T),
+    None,
+}
+
+impl<S, C> Batcher<S, C>
+where
+    S: Stream,
+    C: BatchConfig<S::Item>,
+{
+    pub fn new(stream: S, config: C) -> Self {
+        Self {
+            config,
```

Comment:
Small nit: rename this to `inner` or `state` or just `batch`.

I think it reads weird otherwise.  For example, `if this.config.batch_len() == 0`.  Does that means our batch length is configured to be zero i.e. never batch?